scalableminds/webknossos
# Markdown comments

This PR virtualizes the comment list in the tracing view, enables markdown rendering for comments and adds clickable references to other nodes and positions in comments.
Multiline comments are shown in proper markdown in a small popover next to the comment, when active.
I've also added a modal to enable proper markdown editing in a multiline textarea (click on edit markdown button). The sorting in the comment tab was reworked as well, natural sort is no longer the default (too slow) and it is now possible to reverse the sorting order in all sort modes (not only when sorting by name as previously).
I'm not entirely happy with the UI. The display of very large comments in the popover looks out of place and the single line comment input is also not perfect. One potential problem I see there is that users hit "Enter" to create a new comment, which now creates a new line and makes users believe their comment was not saved (as they only see the new empty line). Any recommendations appreciated :)

@fm3 As mentioned today, the backend will need to rewrite references in comments, that reference other nodes (`#<nodeid>`), in case nodeIds are changed during a merge. The frontend code to do this does a simple regex replace on comments using the regex `/#([0-9]+)/g`. If the referenced id is invalid, not existing or unchanged, nothing should be replaced.
I've also noticed that line breaks are stripped from comments when uploading an NML through the backend, could you check how to not do this :)

### URL of deployed dev instance (used for testing):
- https://markdowncomments.webknossos.xyz

### Steps to test:
- Upload NML with multiline comments and markdown using frontend/backend - The comments should be displayed correctly
- Create multiline comments with markdown using the edit markdown modal
- Reference other nodeIds (`#<nodeid>`) or positions (`#(<x,y,z>)`) which should be clickable and activate the referenced node / set the referenced position on click
- Also test Navigation to Dashboard/Other pages and test task hot-swap (because I removed code in the url_manager that was originally needed for this functionality to work correctly - however, during my testing I didn't notice any problems)

### TODO
 - [ ]  Compatibility btw front-end and back-end XML parser/writer concerning line-breaks

### Issues:
- fixes #2824

------
